extensions/amp-date-display/1.0/date-display.js

```diff
@@ -65,26 +93,56 @@ let EnhancedVariablesV2Def;
  * @param {!DateDisplayDef.Props} props
  * @return {PreactDef.Renderable}
  */
-export function DateDisplay(props) {
-  const {render, children} = props;
-  const data = /** @type {!JsonObject} */ (getDataForTemplate(props));
+export function DateDisplay({
+  datetime,
+  timestampMs,
+  timestampSeconds,
+  displayIn,
+  locale,
+  offsetSeconds,
+  render = DEFAULT_RENDER,
+  ...rest
+}) {
+  const epoch = getEpoch(datetime, timestampMs, timestampSeconds);
+  // TODO(#29293): Move defaults to the props destructuring in the args once
+  // the typing is fixed.
+  const data = useMemo(
+    () =>
+      getDataForTemplate(
+        epoch,
+        displayIn || DEFAULT_DISPLAY_IN,
+        locale || DEFAULT_LOCALE,
+        offsetSeconds || DEFAULT_OFFSET_SECONDS
+      ),
+    [epoch, displayIn, locale, offsetSeconds]
+  );
+
+  const rendered = useRenderer(render, data);
+  const isHtml =
+    rendered && typeof rendered == 'object' && '__html' in rendered;
+
   useResourcesNotify();
 
-  return render(data, children);
+  return (
+    <Wrapper
+      {...rest}
+      as="time"
+      datetime={data['iso']}
+      dangerouslySetInnerHTML={isHtml ? rendered : null}
+    >
+      {isHtml ? null : rendered}
+    </Wrapper>
+  );
 }
 
 /**
- * @param {!DateDisplayDef.Props} props
+ * @param {number|undefined} epoch
+ * @param {string} displayIn
+ * @param {string} locale
+ * @param {number} offsetSeconds
  * @return {!EnhancedVariablesV2Def}
  */
-function getDataForTemplate(props) {
-  const {
-    displayIn = '',
-    locale = DEFAULT_LOCALE,
-    offsetSeconds = DEFAULT_OFFSET_SECONDS,
-  } = props;
-
-  const epoch = getEpoch(props);
+function getDataForTemplate(epoch, displayIn, locale, offsetSeconds) {
```

Comment:
Continuing the thread from yesterday: Default values can still apply even when the parameter is specified, if the value is `undefined`:

```js
function foo(bar = true) {
  return bar;
}

assert.equals(foo(), foo(undefined));
```

Reply:
wow! totally didn't expect that. will update.

Reply:
Sorry, keep confusing myself about this. What I really want is to define defaults in the component's signature like this:

```
function DateDisplay({
  ...
  displayIn: DEFAULT_DISPLAY_IN,
  ...
}) {
  callToSomeMethod(displayIn);  // Here, `displayIn` is assumed to be just `string`, not `string|undefined` anymore.
}
```

And then to use it with the right type in all the downstream code. I really would rather have a single place where this happens for now. Otherwise all the defaults are applied multiple types throughout. This is what the #29293 is all about.

Reply:
I moved the defaults higher up and annotated them a bit clearer. LMK if you think it's good enough for now.

Reply:
Yah, if we move them into the `DateDisplay` param destructure, that's fine with me.

Reply:
Done. But had to remove `|undefined` from types for now. Let's follow up on this in #29293 to keep threads separate.